exist-core/src/main/java/org/exist/xquery/Cardinality.java

```diff
@@ -1,91 +1,184 @@
 /*
- *  eXist Open Source Native XML Database
- *  Copyright (C) 2001-06 Wolfgang M. Meier
- *  wolfgang@exist-db.org
- *  http://exist.sourceforge.net
- *  
- *  This program is free software; you can redistribute it and/or
- *  modify it under the terms of the GNU Lesser General Public License
- *  as published by the Free Software Foundation; either version 2
- *  of the License, or (at your option) any later version.
- *  
- *  This program is distributed in the hope that it will be useful,
- *  but WITHOUT ANY WARRANTY; without even the implied warranty of
- *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
- *  GNU Lesser General Public License for more details.
- *  
- *  You should have received a copy of the GNU Lesser General Public License
- *  along with this program; if not, write to the Free Software
- *  Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
- *  
- *  $Id$
+ * eXist-db Open Source Native XML Database
+ * Copyright (C) 2001 The eXist-db Authors
+ *
+ * info@exist-db.org
+ * http://www.exist-db.org
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
  */
 package org.exist.xquery;
 
+import static org.exist.xquery.Cardinality.InternalValue.*;
+
 /**
- * Defines occurrence indicators (*,?,+).
- * 
- * @author wolf
+ * Defines <a href="https://www.w3.org/TR/xpath-31/#prod-xpath31-OccurrenceIndicator">XPath Occurrence Indicators</a>
+ * (*,?,+), and additionally defines {@link #EMPTY_SEQUENCE}, and {@link #EXACTLY_ONE} for convenience.
+ *
+ * @author <a href="mailto:adam@evolvedbinary.com">Adam Retter</a>
  */
-public class Cardinality {
-
-	public final static int ZERO = 1;	
-	public final static int ONE = 2;	
-	public final static int MANY = 4;	
-	public final static int EMPTY = ZERO;
-	public final static int EXACTLY_ONE = ONE;
-	/** indicator '+' **/
-	public final static int ONE_OR_MORE = ONE | MANY;
-	/** indicator '*' **/
-	public final static int ZERO_OR_MORE = ZERO | ONE | MANY;
-	/** indicator '?' **/
-	public final static int ZERO_OR_ONE = ZERO | ONE;
-    
-    public final static String toString(int cardinality) {
-        switch(cardinality) {
-            case EMPTY:
+public enum Cardinality {
```

Comment:
@adamretter do you have an idea on performance? e.g the GC as there are objects involved iso a simple type.

Reply:
Enum members are statically allocated. So there are only ever as many objects as there are values of the enum. Also everything about the enum is statically known, where as with mutable int flags they can't be determined statically, so actually I would guess the enum allows the compiler to produce more efficient code.